text/2018-10-25-tikv-client-rust.md

```diff
@@ -0,0 +1,480 @@
+
+## Naming
+
+The [Rust API
+Guidelines](https://rust-lang-nursery.github.io/api-guidelines/naming.html) do
+not prescribe any particular crate name convention.
+
+We choose to name the crate `tikv-client`. Choosing to seperate `tikv` and
+`client` helps potentially unfamiliar users to immediately understand the
+intent of the package. `tikvclient`, while understandable, is not immediately
+parsable by a human.
+
+All structures and functions will otherwise follow the [Rust API
+Guidelines](https://rust-lang-nursery.github.io/api-guidelines/), some of which
+will be enforced by `clippy`.
+
+## Installation
+
+To utilize the client programmatically, users will be able to add the
+`tikv-client` crate to their `Cargo.toml`'s dependencies. Then they must use
+the crate with `use tikv_client;`. Unfortunately due to Rust’s naming
+conventions this inconsistency is in place.
+
+## Usage
+
+## Two types of APIs
+
+TiKV provides two types of APIs for developers:
+- The Raw Key-Value API
+
+    If your application scenario does not need distributed transactions or MVCC
+(Multi-Version Concurrency Control) and only needs to guarantee the atomicity
+towards one key, you can use the Raw Key-Value API.
+
+- The Transactional Key-Value API
+
+    If your application scenario requires distributed ACID transactions and the
+atomicity of multiple keys within a transaction, you can use the Transactional
+Key-Value API.
+
+Generally, the Raw Key-Value API has higher throughput and lower latency
+compared to the Transactional Key-Value API. If distributed ACID transactions
+are not required, Raw Key-Value API is preferred over Transactional Key-Value
+API for better performance and ease of use. **Please be aware that these two
+types of APIs are mutually exclusive. Users should __never__ mix and use these
+two types of APIs on a __same__ TiKV cluster.**
+
+The client provides two types of APIs in two separate modules for developers to
+choose from.
+
+### The common data types
+
+- `Key`: raw binary data
+- `Value`: raw binary data
+- `KvPair`: Key-value pair type
+- `KvFuture`: A specialized Future type for TiKV client
+- `Config`: Configuration for client
+
+### Raw Key-Value API Basic Usage
+
+To use the Raw Key-Value API, take the following steps:
+
+1. Create an instance of Config to specify endpoints of PD (Placement Driver)
+and optional security config
+
+    ```rust
+    let config = Config::new(vec!["127.0.0.1:3379"]).with_security(
+        PathBuf::from("/path/to/ca.pem"),
+        PathBuf::from("/path/to/client.pem"),
+        PathBuf::from("/path/to/client-key.pem"),
+    );
+    ```
+
+2. Create a Raw Key-Value client.
+
+    ```rust
+    let client = raw::Client::new(&config);
+    ```
+
+3. Call the Raw Key-Value client methods to access the data on TiKV. The Raw
+Key-Value API contains following methods
+
+    ```rust
+    impl Client {
+        pub fn new(_config: &Config) -> Connect;
```

Comment:
One more question about `Config`: Is it certern that `Connect` takes the ownership of a `Config`? According to [the API Guidelines](https://rust-lang-nursery.github.io/api-guidelines/flexibility.html#caller-decides-where-to-copy-and-place-data-c-caller-control), if we use the config as owned here, a `Config` should be taken instead of a reference.

Reply:
It's not decided yet. We'll see

Reply:
See https://github.com/tikv/client-rust/pull/33